staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -594,6 +599,35 @@ func (w *watchCache) GetAllEventsSinceThreadUnsafe(resourceVersion uint64) ([]*w
 		// current state and only then start watching from that point.
 		//
 		// TODO: In v2 api, we should stop returning the current state - #13969.
+		ci := &watchCacheInterval{fillAll: true}
+		return ci, nil
+	}
+	if resourceVersion < oldest-1 {
+		return nil, errors.NewResourceExpired(fmt.Sprintf("too old resource version: %d (%d)", resourceVersion, oldest-1))
+	}
+
+	// Binary search the smallest index at which resourceVersion is greater than the given one.
+	f := func(i int) bool {
+		return w.cache[(w.startIndex+i)%w.capacity].ResourceVersion > resourceVersion
+	}
+	first := sort.Search(size, f)
+	ci := &watchCacheInterval{
+		startIndex: (w.startIndex + first),
+		endIndex:   (w.endIndex - 1),
+		indexer: func(i int) *watchCacheEvent {
+			return w.cache[i%w.capacity]
+		},
+	}
+	return ci, nil
+}
+
+func (w *watchCache) GetAllEventsSince(resourceVersion uint64) ([]*watchCacheEvent, error) {
+	cacheInterval, err := w.getCacheIntervalForEvents(resourceVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	if cacheInterval.fillAll {
```

Comment:
I'm not a fan of splitting the actual logic into two methods - let's leave the logic inside thread-unsafe function.